rules/prefer-error-cause.js

```diff
@@ -0,0 +1,392 @@
+'use strict';
+const _ = require('lodash');
+const esquery = require('esquery');
+const {isCommaToken} = require('eslint-utils');
+const {appendArgument} = require('./fix/index.js');
+const {methodCallSelector} = require('./selectors/index.js');
+const getReferences = require('./utils/get-references.js');
+const getMatchingAncestorOfType = require('./utils/get-matching-ancestor-of-type.js');
+
+const ERROR = 'error';
+
+const messages = {
+	[ERROR]: 'Prefer specifying an old error as `cause` option when rethrowing the error.',
+};
+
+const promiseThenSelector = methodCallSelector({method: 'then', argumentsLength: 2});
+const promiseCatchSelector = methodCallSelector({method: 'catch', argumentsLength: 1});
+
+const isThenMethod = callExpression =>
+	callExpression?.optional !== true
+	&& callExpression?.arguments?.length === 2
+	&& callExpression?.arguments[0].type !== 'SpreadElement'
+	&& callExpression?.arguments[1].type !== 'SpreadElement'
+	&& callExpression?.callee.type === 'MemberExpression'
+	&& callExpression?.callee.optional !== true
+	&& callExpression?.callee.computed !== true
+	&& callExpression?.callee.property.type === 'Identifier'
+	&& callExpression?.callee.property.name === 'then';
+
+const isCatchMethod = callExpression =>
+	callExpression?.optional !== true
+	&& callExpression?.arguments?.length === 1
+	&& callExpression?.arguments[0].type !== 'SpreadElement'
+	&& callExpression?.callee.type === 'MemberExpression'
+	&& callExpression?.callee.optional !== true
+	&& callExpression?.callee.computed !== true
+	&& callExpression?.callee.property.type === 'Identifier'
+	&& callExpression?.callee.property.name === 'catch';
+
+const isFunctionType = type => type === 'FunctionExpression' || type === 'ArrowFunctionExpression' || type === 'FunctionDeclaration';
+
+const isArgumentOfFunction = (identifier, {rootNode, functionType}) => {
+	const selector = functionType === 'then' ? promiseThenSelector : promiseCatchSelector;
+	const argumentOrder = functionType === 'then' ? 1 : 0;
+	const selectedExpressions = esquery.match(rootNode, esquery.parse(selector));
+
+	for (const expression of selectedExpressions) {
+		if (
+			expression.arguments[argumentOrder].type === 'Identifier'
+			&& expression.arguments[argumentOrder].name === identifier
+		) {
+			return true;
+		}
+
+		if (
+			isFunctionType(expression.arguments[argumentOrder].type)
+			&& esquery.matches(expression.arguments[argumentOrder].body, esquery.parse(`[type="CallExpression"][callee.name="${identifier}"]`))
+		) {
+			return true;
+		}
+	}
+
+	return false;
+};
+
+const getCatchBlock = node => {
```

Comment:
I have trouble to understand this function, can we do this.

1. Add a function to search ancestor that is function or `CatchClause`, whatever comes first.
2. Add a function to test that ancestor is `.then`/`.catch`.

The "CatchBlock" should always be a function or `CatchClause`.

Correct me if this logic is wrong.

Reply:
> 1. Add a function to search ancestor that is function or CatchClause, whatever comes first.

I think `getCatchBlock` is doing this although it additionally checks the file scope's callback function.

> 2. Add a function to test that ancestor is .then/.catch's argument.

I added `isThenOrCatchMethod` method for this thing and to simplify the code.

> The "CatchBlock" should always be a function or CatchClause, and it should not cross function or CatchClause.

I think `getCatckBlock` should return `function` or `CatchClause` if my implementation is not wrong.